### PR TITLE
GEODE-8200: enhance GfshRule to specify a working dir

### PR DESCRIPTION
**Pull request by jinmeiliao**
* improve some backward compatibility test to cover more versions.

### CODE REVIEW

File: geode-assembly/src/upgradeTest/java/org/apache/geode/management/DeploymentManagementUpgradeTest.java [hunk 1]
```diff
@@ -55,16 +84,16 @@ public static void beforeClass() throws Exception {
...
   public void newLocatorCanReadOldConfigurationData() throws IOException {
-    File workingDir = tempFolder.newFolder();
     int[] ports = AvailablePortHelper.getRandomAvailableTCPPorts(3);
-    oldGfsh.execute("start locator --name=test --port=" + ports[0] + " --http-service-port="
-        + ports[1] + " --dir=" + workingDir.getAbsolutePath() + " --J=-Dgemfire.jmx-manager-port="
-        + ports[2],
-        "deploy --jar=" + clusterJar.getAbsolutePath(),
-        "shutdown --include-locators");
+    GfshExecution execute =
+        GfshScript.of(startLocatorCommand("test", ports[0], ports[2], ports[1], 0))
```
Comment (jchen21) [inline geode-assembly/src/upgradeTest/java/org/apache/geode/management/DeploymentManagementUpgradeTest.java:89]: For code readability, it is better to have some meaningful names for the ports. Before the code change, it is easy to identify the purpose of a port with the gfsh option in the code. Now it is not that straight forward to tell ports[0] is locator port. port[2] is JMX port and ports[1] is HTTP port, unless you see to the definition of `startLocatorCommand`.

File: geode-junit/src/main/java/org/apache/geode/test/junit/rules/gfsh/GfshRule.java [hunk 2]
```diff
@@ -199,4 +220,23 @@ private void stopMembers(GfshExecution gfshExecution) {
     }
     execute(GfshScript.of(stopMemberScripts).withName("Stop-Members"));
   }
+
+  public static String startServerCommand(String name, int port, int connectedLocatorPort) {
```
Comment (kirklund) [inline geode-junit/src/main/java/org/apache/geode/test/junit/rules/gfsh/GfshRule.java:224]: These new static methods don't really fit well in GfshRule. I think you should probably move them to two new classes (probably not a rule) named something like ServerCommandBuilder and LocatorCommandBuilder, and explode the parameters to each have their own setter type method:
```
String startServerCommand = new ServerCommandBuilder()
    .withPort(port)
    .withLocator(locatorPort)
    .create(name);
```

Comment (kirklund) [reply geode-junit/src/main/java/org/apache/geode/test/junit/rules/gfsh/GfshRule.java:224]: There are actually a couple existing classes like this that we could move to geode-dunit and make any necessary changes:
```
geode-core/src/integrationTest/java/org/apache/geode/distributed/LocatorCommand.java
geode-core/src/integrationTest/java/org/apache/geode/distributed/ServerCommand.java
```

Comment (kirklund) [reply geode-junit/src/main/java/org/apache/geode/test/junit/rules/gfsh/GfshRule.java:224]: We could commit what you have ready and then make further changes. I'll go ahead and approve and let you decide.

Comment (jinmeiliao) [reply geode-junit/src/main/java/org/apache/geode/test/junit/rules/gfsh/GfshRule.java:224]: Thanks! I will merge this now and we can make improvements on it when we get to consolidating all these methods into one place.

**Pull request merged**